### ISSUE TITLE
memory leak outside vm ?

### ISSUE THREAD

**Issue by fvigotti** [kubernetes-client/java#1002]
I have a simple program in java that watch the kubernetes cluster,
Just update the client from verison 3+ to 6.0.1
the only call the program does is this : 
```
api.listNamespacedPod(k8sNamespace, "false",null,"",labels,10,null,10,false)
```
about every 3minutes.. 
now I'm experiencing memory leak, and pods gets killed ( 300mb Xmx , no problems with garbage collection which keep memory usage inside the jvm to about 120mb ) but the leaks is outside the jvm ( only 1 container in that pod with just this jvm as the only running program ) which even with a limit of 1200mb, after some hours crashes.. this in all nodes and each 5/6 hours 

I suspect that this library have a memory leak ( this is the only update the in program in 1.5 years.. never experienced problems and not OOM every 5/6 hours )

**Comment by brendandburns** [comment]
The only way that this code could leak memory that wasn't inside the JVM would be if it was leaking code from native code. I don't think that code path uses any native code except the lowest level JVM socket code which I'd be surprised if it was leaking, especially if you haven't updated the JVM.

given your use-case though, I suspect that you would be better off using a watch:
https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes

https://github.com/kubernetes-client/java/blob/master/examples/src/main/java/io/kubernetes/client/examples/InformerExample.java

Rather than polling every 3 minutes.

**Comment by fvigotti** [comment]
I've just downgraded the library to 5.0.0, and now everything seems fine..

yeah jvm versions didn't changed , It's the bit outdated v8u172

but given that downgrading the problem disappeared... I strongly suspect there is something wrong in that 6.0.1 release
unfortunately I have no time now to debug further.. but maybe this post could help others in same situation

thank you,
Francesco

**Comment by brendandburns** [comment]
Closing since it seems resolved.

**Comment by anupreet5** [comment]
It seems like this issue is still not resolved.
I'm facing the same problem when my service is running inside the kubernetes environment and the pod automatically gets killed after some time (OOMKilled) with the usage of "listNamespacedPod". This is with version 10.0.0.

However as advised by Francesco, I moved to version 5.0.0 and the issue is no more seen.
I think some resolution should be done in the newer versions.

**Comment by yue9944882** [comment]
@fvigotti @anupreet5 can you share me how to reproduce the leakage? alternatively you can move to the informer to watch the changes.. it's proven in our production at least..

**Comment by anupreet5** [comment]
For recreating the issue you can create a simple spring boot scheduler with a high frequency and call the api mentioned in 1st comment (or alternatively you can also put it in for loop).
Run this application in a kubernetes cluster, do analyze the memory while running this, after some time OOM!

Sample code:

`
  @PostConstruct
  private void init() throws IOException {
    ApiClient client = ClientBuilder.cluster().build();
    Configuration.setDefaultApiClient(client);
  }

  @Scheduled(fixedRateString = "500", initialDelayString = "2000")
  public void refreshKubernetes() throws ApiException, IOException {

    CoreV1Api api = new CoreV1Api();
    V1PodList pods = api.listNamespacedPod(healthProperties.getKubernetesNamespace(), true, null,
        null, null, null, 100, null, 10, null);
    // List<String> kubernetesServiceNames = some services name

    for (String kubernetesService : kubernetesServiceNames) {
      long totalCount = pods.getItems().stream()
          .filter(pod -> pod.getMetadata().getName().contains(kubernetesService))
          .count();
      long activeCount = pods.getItems().stream()
          .filter(pod -> pod.getMetadata().getName().contains(kubernetesService))
          .filter(pod -> pod.getStatus().getContainerStatuses().get(0).isReady())
          .count();
      
      log.info("Service status: {}", totalCount - activeCount);
  }
`

**Comment by anupreet5** [comment]
Just an update, I saw the issue even in 5.0.0
I analyzed the heap dump and I suspect it is caused by "com/squareup/okhttp/ConnectionPool".

@yue9944882 Can you please check?

**Comment by brendandburns** [comment]
I think it is because you are creating a new HTTP Client every time your "refreshKubernetes()" function is called, since you are instantiating `CoreV1Api` in that function. You should instantiate that once outside of the function and just use it in your function.

**Comment by anupreet5** [comment]
I did verify that as well, however still the same behavior.
Also I checked the CoreV1Api constructor and felt like it doesnt do any big work, so kept it inside.

Anyway I had a chance to debug more on this.
It seems like there's isnt any memory leak from the code perspective, however for some reason the garbage collection is not happening. I observed the memory consumption through VisualVM, manually doing GC does free up the space in heap.

I'm not sure how to attach a snapshot here. But the heap consumption went more than >1GB, and still GC didnt happen.
For now, I have solved the issue by using these jvm command line options.
**-Xms512m -Xmx1024m**

**Comment by yue9944882** [comment]
@fvigotti @anupreet5 are you constructing multiple `ApiClient` instances in your code? one significant difference between 5.0.0 and 6.0.1 is that it bumps okhttp dependency from 2.x.x -> 3.x.x. i noticed thread leakage when we landed 6.0.1 in the first, and i believe that's caused by the following code from okhttp library where the [clean-up thread-pool](https://github.com/square/okhttp/blob/06644bb0507873e9a3b89d9107da537f1b140e91/okhttp/src/main/java/okhttp3/internal/connection/RealConnectionPool.java#L46) is private static but the [clean-up runnable](https://github.com/square/okhttp/blob/06644bb0507873e9a3b89d9107da537f1b140e91/okhttp/src/main/java/okhttp3/internal/connection/RealConnectionPool.java#L53-L68) is pushed into the thread pool every time we add a connection to the pool. everything will be fine if the clean-up runnable does a one-time clean-up but unfortunately it keeps cleaning up in a while(true) condition which may protentially cause the runnables lumping in the thread-pool.


https://github.com/square/okhttp/blob/06644bb0507873e9a3b89d9107da537f1b140e91/okhttp/src/main/java/okhttp3/internal/connection/RealConnectionPool.java#L120

one way to solve this is to bump okhttp to 4.x.x i suppose.

**Comment by anupreet5** [comment]
No no, I'm instantiating it just once.

As stated in my previous comment, I am no more seeing the issue (with 5.0.0 as well as 10.0.0) by using those vm args.
Those will make sure that GC happens if the memory reaches near the limit.
Ideally without those args, GC should be done when it is going OOM, but for some reason it doesn't.

Anyone facing this issue can try setting those flags according to their application.
_p.s. Having low threshold for -Xms will cause frequent GC._
